datadriven/src/test_data_reader.rs

```diff
@@ -0,0 +1,159 @@
+            }
+
+            debug!(self.logger, "cmd: {}, cmd_args: {:?}", cmd, cmd_args,);
+
+            self.data.cmd = cmd;
+            self.data.cmd_args = cmd_args;
+
+            loop {
+                let line = self.scanner.next();
+                if line.is_none() {
+                    bail!("testdata is not complete, we expected '----' after command");
+                }
+
+                let line = line.unwrap().1;
+
+                if line == "----" {
+                    self.read_expected();
+                    return Ok(true);
+                } else if !line.is_empty() {
+                    bail!("expected '----', found '{}'", line);
+                }
+            }
+        }
+    }
+
+    // check whether there is two separator or one
+    fn read_expected(&mut self) {
+        // after reading first separator
+        // we expected two input
+        // (1) second separator
+        // (2) non empty output
+
+        // We need to add an end line for each line in order to distinguish the exact output,
+        // otherwise we may consider an unexpected output combination to be correct
+
+        if let Some((_, line)) = self.scanner.next() {
+            if line == "----" {
+                loop {
+                    let mut line = self.scanner.next().unwrap().1.trim().to_string();
+                    if line == "----" {
+                        let mut line2 = self.scanner.next().unwrap().1.trim().to_string();
+                        if line2 == "----" {
+                            let line3 = self.scanner.next().unwrap().1.trim().to_string();
+                            assert!(
+                                line3.is_empty(),
+                                "we expected an empty line after second separator, found '{}'",
+                                line3
+                            );
+                            break;
+                        }
+                        if !line2.is_empty() {
+                            writeln!(&mut line2).unwrap();
+                            self.data.expected.push_str(line2.as_str());
+                        }
+                    }
+                    if !line.is_empty() {
+                        writeln!(&mut line).unwrap();
+                        self.data.expected.push_str(line.as_str());
+                    }
+                }
+            } else {
+                let mut l = line.trim().to_string();
+                if !l.is_empty() {
+                    writeln!(&mut l).unwrap();
+                    self.data.expected.push_str(l.as_str());
+                }
+                loop {
+                    let mut line = self.scanner.next().unwrap().1.trim().to_string();
+                    if line.is_empty() {
+                        break;
+                    } else {
+                        writeln!(&mut line).unwrap();
+                        self.data.expected.push_str(line.as_str());
+                    }
+                }
```

Comment:
To keep consisitent with the original go implementation. Maybe we can use a similar approach like this?
```rust
loop {
   if l.is_empty() {
        break;
   }
   writeln!(&mut l).unwrap();
   self.data.expected.push_str(l.as_str());
   l = self.scanner.next().unwrap().1.trim().to_string();
}
```